woocommerce-services.php

```diff
@@ -339,9 +339,23 @@ function() {
 				}
 			);
 			add_action( 'plugins_loaded', array( $this, 'jetpack_on_plugins_loaded' ), 1 );
+			add_action( 'plugins_loaded', array( $this, 'maybe_handle_coexistence_with_woo_shipping_and_woo_tax' ), 5 );
 			add_action( 'plugins_loaded', array( $this, 'on_plugins_loaded' ) );
 		}
 
+		/**
+		 * Used to let Woo Shipping and Woo Tax know WCS&T will handle the plugins' coexistence
+		 * by displaying an appropriate notice and not registering its functionality.
+		 */
+		public function maybe_handle_coexistence_with_woo_shipping_and_woo_tax() {
+			$is_woo_shipping_active = in_array( 'woocommerce-shipping/woocommerce-shipping.php', get_option( 'active_plugins' ) );
+			$is_woo_tax_active      = in_array( 'woocommerce-tax/woocommerce-tax.php', get_option( 'active_plugins' ) );
+
+			if ( $is_woo_shipping_active && $is_woo_tax_active ) {
+				add_filter( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', '__return_true' );
+			}
```

Comment:
[Subjective, non-blocking] Could we potentially simplify the solution to "just" return if Woo Shipping and Woo Tax is activated and then call the method directly when we call `apply_filters`?

```php
apply_filters( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', $this->maybe_handle_coexistence_with_woo_shipping_and_woo_tax() );
```

I see a few potential benefits:

1. It's very minor, but it is a slight performance increase since we can get rid of 1 action call (`plugins_loaded`) and 1 filter (`wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax`).
2. By not adding any `add_filter( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', ... );` ourselves, we reduce the chance of race conditions to happen if others hook into the filter.
    * _This argument is **super** theoretical since we do not expect others to use our hooks._
3. Opinionated, then it makes the implementation simpler and therefore easier to read and maintain for other developers.

That way we achieve  is still used as the fallback value and will work exactly like now, but we don't have to use a `add_filter` ourselves that is being executed in an additional `plugins_loaded` action? 😄  

```suggestion
			return $is_woo_shipping_active && $is_woo_tax_active;
```

Reply:
> It's very minor, but it is a slight performance increase since we can get rid of 1 action call (plugins_loaded) and 1 filter (wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax).

Hey, you're right! Since we're reading the option and not actually waiting for plugins to load, we can do the check as soon as `get_option()` is available.

> By not adding any add_filter( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', ... ); ourselves, we reduce the chance of race conditions to happen if others hook into the filter. 

I'm not sure I understand. What is the difference if we don't send any value to the hook?

> Opinionated, then it makes the implementation simpler and therefore easier to read and maintain for other developers.

Yes! Very cool.

Reply:
> That way we still use the same logic as as the fallback value for our filter and it'll work exactly like now, but we don't have to use a add_filter ourselves that is being executed in an additional plugins_loaded action? 😄

Oh, wait; I committed the change you suggested but we do have to register the `add_filter` call. Otherwise Woo Shipping and Woo Tax won't get the message.

Reply:
Nice! That all sounds awesome! 😄 

>I'm not sure I understand. What is the difference if we don't send any value to the hook?

What I meant was that instead of having both a place where "call" the hook and registering our own as well:

```php
// __construct():
if ( $this->are_woo_shipping_and_woo_tax_active() ) {
	/**
	 * Used to let Woo Shipping and Woo Tax know WCS&T will handle the plugins' coexistence
	 * by displaying an appropriate notice and not registering its functionality.
	 */
	add_filter( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', '__return_true' );
	return;
}

// on_plugins_loaded():
apply_filters( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', false )
```

It might be possible to combine them into this:

```php
apply_filters( 'wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax', $this->are_woo_shipping_and_woo_tax_active() );
```

The primary (very theoretical) upside is that we reduce the chance for race conditions with other places that wants to use the same hook.
It's really not important since I do not expect 3. party developers to add hooks as well, which is why it's super theoretical 😄 

The second benefit is just a small performance improvement since and - depending on your pov - a simplified codebase

---

> Oh, wait; I committed the change you suggested but we do have to register the add_filter call. Otherwise Woo Shipping and Woo Tax won't get the message.

Ah, sorry! It was meant as a partial example since I couldn't modify the code in multiple places 🙈